numpy/lib/function_base.py

```diff
@@ -175,54 +175,114 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
             raise AttributeError(
                 'max must be larger than min in range parameter.')
 
+    # Histogram is an integer or a float array depending on the weights.
+    if weights is None:
+        ntype = np.dtype(np.intp)
+    else:
+        ntype = weights.dtype
+
+    # We set a block size, as this allows us to iterate over chunks when
+    # computing histograms, to minimize memory usage.
+    BLOCK = 65536
+
     if not iterable(bins):
+
         if np.isscalar(bins) and bins < 1:
             raise ValueError(
                 '`bins` should be a positive integer.')
+
         if range is None:
             if a.size == 0:
                 # handle empty arrays. Can't determine range, so use 0-1.
                 range = (0, 1)
             else:
                 range = (a.min(), a.max())
+
         mn, mx = [mi + 0.0 for mi in range]
+
         if mn == mx:
             mn -= 0.5
             mx += 0.5
+
+        # At this point, if the weights are not integer, floating point, or
+        # complex, we have to use the slow algorithm.
+        if weights is not None and not weights.dtype.kind in ('i', 'f', 'c'):
```

Comment:
I think this would be better implemented as:

```
if weights is not None and not (np.can_cast(weights.dtype, np.double) or
                                np.can_cast(weights.dtype, np.complex)):
```